### ISSUE TITLE
Ability to mark user defined documents/rules/functions as deprecated

### ISSUE THREAD

**Issue by johnrengelman** [open-policy-agent/opa#6077]
## What is the underlying problem you're trying to solve?
We are building a standardized auth system for multiple service teams to use using OPA. As we are iterating and releasing new capabilities, we've found that as we adopt changing requirements, we want to be able to deprecate common rules or functions with improved options and not necessarily maintain backwards compatibility. Currently, the only way to do this is to use a _custom_ METADATA attribute to indicate the deprecation. However, this is documentation only and doesn't integrate into the tooling in anyway.

## Describe the ideal solution
Introduce a formal `deprecated: <boolean>` METADATA attribute and integrate it with the `opa check` (perhaps limit it to `--strict`) such that usage of deprecated rules/functions is indicated/error'd.

For deprecated documents, an additional message can be returned in the response to indicate the deprecation:
```
{"result": true, "deprecated": "the document x/y/z is deprecated"}
```

For referencing documents or functions in other rules or functions, `opa check` should output errors and locations of the usages.
Ideally, deprecation detections could be suppressed or ignored either via another METADATA attribute at the usage point or via an external file.

## Describe a "Good Enough" solution
Starting by just having `opa check` output usage of deprecated documents or functions would be a good start. If the output was machine-formatted, we can write our own ignore capability by processing the output.

**Comment by ashutosh-narkar** [comment]
This seems like a good feature. Surfacing this when strict mode is enabled makes sense. This should also work well with the scoping mechanism. @johanfylling wdyt?

**Comment by anderseknert** [comment]
This would be a fairly simple addition to [Regal](https://github.com/StyraInc/regal) using what's currently available, i.e. a `deprecated` attribute under the `custom` map. I do agree that this would be nice to support in OPA proper though! I wonder if we should limit the attribute to a boolean value, or if we'd also want to support an object form (or whatever) where something like "since" could be expressed? Tooling like `opa check` wouldn't need to care, but it would be nice for documentation purposes.

> For deprecated documents, an additional message can be returned in the response to indicate the deprecation

I wonder if there are clients that would break if we introduce new, unknown (to them) attributes in the response? 🤔 I mean, I would _hope_ most clients are configured to ignore additional attributes, but who knows. Perhaps it would be good to play it safe, and only include this if e.g. some query parameter in the request signaled that this should be included in the response? Similarly to how we do this for metrics, and whatnot.

**Comment by anderseknert** [comment]
A response header would work, but I think I'd prefer to include it in the body like your original suggestion, but only provided that some query param was included with the request, like `?check=true` (or whatever), as that would align better with how we deal with other optional things in the response payload, like metrics. See https://www.openpolicyagent.org/docs/latest/rest-api/#query-parameters-4

That would also provide consistency for non-server contexts like `opa eval`, where we could add a new flag to have these warnings included in the result as well.

If we do add `since`, I guess we'd need to add a top level `since` attribute too, or else it would be hard to know what it refered to. If we're creeping to far out of scope, we could just start with a boolean attribute and consider more options later.

**Comment by johanfylling** [comment]
Agreed, this is a good feature, and toggling it with `strict-mode` is a good mechanism for keeping things tidy 👍.

Paired with Regal, I could even imagine the very first iteration of this just being the METADATA update, and not doing anything fancy in the compiler. In my mind this lives pretty firmly in linter-land 😉 .

To contradict my above statement, though: given that rules, etc. can be dynamically referenced, I think it'd be difficult to statically catch all cases, though. Collecting warnings at eval-time is probably the best way to catch trickier edge cases where policies are very dynamic. And even then, we'd only catch cases as they happen, and rare usage might go undetected for quite some time.

**Comment by anderseknert** [comment]
Good call on dynamic function calls. @johanfylling! 

Although I don't think we support those fully in OPA today, there's no reason (that I can think of) we shouldn't, and I think there's an open issue or two in the backlog on that topic as well. 

Probably a good idea to separate, and expand on, the expected compile/build time behavior (i.e. `opa check` and friends) from the expected runtime behavior (`opa run`, `opa eval`...) a bit further here.
